[Step3] 자동차 경주

namonak opened the pull request.

namonak committed:
refactor(step2): change the Operand class to a data class

namonak committed:
feat: add a Position class that stores the position of the car

namonak committed:
feat: add a Car class

The following have been implement and test:
 - add the Position property
 - change the position value according to the input value

namonak committed:
feat: add a Utils class

The following have been implement and test:
 - implement a method that generates a random number by passing the start index and end index as parameters
 - if the start index is bigger than the end index, raise an IllegalArgumentException

namonak committed:
feat: add a Cars class

The following have been implement and test:
 - throw IllegalArgumentException if the input value is less than 1
 - add a play() method that receives a random value and moves the car
 - add a getPositions() method that returns the result

namonak committed:
feat: add InputView and ResultView classes for handling user input and printing result values

namonak committed:
feat: add a RacingGame class that runs a car racing game

namonak committed:
refactor: change the package name according to the class characteristics

namonak committed:
refactor(step2): add a secondary constructor to the Operand class to handle string input

The following have been implement and test:
 - throw an NumberFormatException when an empty string or a non-numeric string is input

namonak committed:
refactor(step2): update the Calculator and StringCalculator classes

The following have been implement and test:
 - wrap operand values in Operand class
 - ensure that the operand values of the Operand class are positive
 - remove java.util.StringTokenizer

sah3122 commented on src/main/kotlin/step3/view/InputView.kt:
```diff
@@ -0,0 +1,30 @@
+        private const val PLAY_TIME_INPUT_MESSAGE = "시도할 횟수는 몇 회인가요?"
+        private const val ERROR = "[ERROR] "
+
+        fun setInputData(): Pair<Int, Int> {
+            return try {
+                val carNumber = inputCarNumber()
+                val playTime = inputPlayTime()
+                Pair(carNumber, playTime)
+            } catch (e: Exception) {
+                println(ERROR + e.message)
+                setInputData()
+            }
+        }
```
자동차 경주에 필요한 입력값을 받는 함수에서는 setter 네이밍을 사용하지 않아도 될것 같아요. 
추가로 Pair를 사용하는것도 좋지만, 입력값을 담고 있는 클래스를 정의하는것도 좋은 방법이 될것 같습니다 ㅎㅎ

namonak replied:
[[10분 테코톡] 📍인비의 DTO vs VO](https://www.youtube.com/watch?v=z5fUkck_RZM) 영상을 참고하여 아래와 같이 수정해 보았습니다. 😃 

- 입력값을 전달하기 위한 `InputResultDto` 클래스 정의
  - 자동차 대수 입력을 위한 `NumberOfCars` 클래스 정의
  - 게임 횟수 입력을 위한 `PlayTime` 클래스 정의

그리고, `setinputData()` 메서드명을 `inputData()` 으로 변경하였습니다!  메서드 네이밍은 항상 어렵네요 😭 

- https://github.com/next-step/kotlin-racingcar/commit/5a33b9b9003a99b82d2cf85bdbc5a8ded913d967

sah3122 commented on src/main/kotlin/step3/view/InputView.kt:
```diff
@@ -0,0 +1,30 @@
+
+        private fun inputCarNumber(): Int {
+            println(CAR_NUMBER_INPUT_MESSAGE)
+            return readLine()!!.toInt()
```
```suggestion
            return readln().toInt()
```
코틀린 버전 1.6 부터 not null 입력값을 받을수 있는 함수가 제공되고 있어요 😄

namonak replied:
꿀팁 감사합니다 😃 아래와 같이 변경해 보았어요! 😄 

- https://github.com/next-step/kotlin-racingcar/commit/ba00a88bd3d8c3d9cec785c04565f2372735b8f9

sah3122 commented on src/main/kotlin/step3/RacingGame.kt:
```diff
@@ -0,0 +1,21 @@
+            repeat(inputData.second) {
+                cars.play()
+                ResultView.printResult(cars)
+            }
```
main 함수를 읽을때 inputData.first / second가 어떤 값을 뜻하는지 나타내질 못하고 있어요 ㅎㅎ
다른 개발자들이 코드를 읽는 경우도 고려해보며 미션을 진행해보세요 !

namonak replied:
조언 감사합니다 😃 
입력값을 `Pair` 객체에 담아 처리하는 경우, 코멘트 남겨주신 것처럼 알아보기 어려운 코드가 되었네요 😭 
`InputResultDto` 클래스를 정의하여 입력값을 처리하였습니다 😃 

- https://github.com/next-step/kotlin-racingcar/commit/5a33b9b9003a99b82d2cf85bdbc5a8ded913d967

sah3122 commented on src/main/kotlin/step3/domain/car/Car.kt:
```diff
@@ -0,0 +1,13 @@
+
+private const val MOVING_MINIMUM_CONDITION_NUMBER = 4
+
+data class Car(val position: Position = Position(0)) {
```
data class 정의 및 Position 정의 훌륭합니다 👍

namonak replied:
칭찬 감사합니다! 😄

sah3122 commented on src/main/kotlin/step3/domain/car/Car.kt:
```diff
@@ -0,0 +1,13 @@
+
+data class Car(val position: Position = Position(0)) {
+    fun move(randomNumber: Int) {
+        if (randomNumber >= MOVING_MINIMUM_CONDITION_NUMBER) {
```
MOVING_MINIMUM_CONDITION_NUMBER 값이 자동차 클래스와 강하게 결합되어 있는데, 이러한 결합도를 낮추기 위해 어떤 노력을 해야할까요 ?

namonak replied:
https://github.com/next-step/kotlin-racingcar/pull/831#discussion_r1020724928 코멘트로 전달해 주신 링크 자료 참조하여, `MovingStrategy` 인터페이스를 추가하였습니다 😃 

이런 방식으로 리팩토링하니 `Car` 클래스에 결합되어 있던 이동 전략에 대한 결합도가 낮추어지는 경험을 하였습니다! 😄

sah3122 commented on src/main/kotlin/step3/domain/car/Cars.kt:
```diff
@@ -0,0 +1,20 @@
+import step3.domain.car.status.Position
+import step3.util.Utils
+
+data class Cars(val car: List<Car>) {
```
자동차 일급 컬렉션 정의 👍 
```suggestion
data class Cars(val cars: List<Car>) {
```
컬렉션 타입은 항상 복수형으로 정의하는것을 권장 드려요.

namonak replied:
네, 코멘트 남겨주신 대로 수정하였습니다! 😃 - https://github.com/next-step/kotlin-racingcar/commit/19f62e5c3341b6d0f40441a1c129087530f0533d

sah3122 commented on src/main/kotlin/step3/domain/car/Cars.kt:
```diff
@@ -0,0 +1,20 @@
+import step3.util.Utils
+
+data class Cars(val car: List<Car>) {
+    private val cars = car
```
해당 클래스는 프로퍼티로 
public val car 와 private val cars 를 가지고 있네요. 
이렇게 정의해주신 이유가 있을까요 ?

namonak replied:
별다른 이유가 없고 제 실수였습니다 😭 다음과 같이 수정하였습니다 😃 

- https://github.com/next-step/kotlin-racingcar/commit/19f62e5c3341b6d0f40441a1c129087530f0533d

sah3122 commented on src/main/kotlin/step3/domain/car/Cars.kt:
```diff
@@ -0,0 +1,20 @@
+    }
+
+    fun play() {
+        cars.forEach { it.move(Utils.generateRandomNumber(Utils.RANDOM_NUMBER_START_INDEX, Utils.RANDOM_NUMBER_END_INDEX)) }
```
자동차의 움직임 전략이 변경되는경우 어떤 문제점이 있을까요 ?
추가로 현재 cars의 play 를 테스트 하기 위해서 Random 함수와 강한 결합이 되어 있습니다. 
이러한 결합을 낮추기 위해 전략 패턴을 고려해볼수 있을것 같아요. 
https://victorydntmd.tistory.com/292

namonak replied:
전달해 주신 링크 자료를 참조하여, `MovingStrategy` 인터페이스를 추가하였습니다. 😃 - https://github.com/next-step/kotlin-racingcar/commit/3f91d198d93dec4d5c6fb77e983b1457bcb85da3

이렇게 리팩토링을 해보니, `CarTest` 테스트코드에서  `move()` 메서드를 테스트할때, `MovingStrategy` 인터페이스의 `isMovable()` 메서드를 오버라이드하여 더욱 직관적으로 테스트 코드를 작성할 수 있었습니다. 😄

sah3122 commented on src/main/kotlin/step3/domain/car/status/Position.kt:
```diff
@@ -0,0 +1,7 @@
+package step3.domain.car.status
+
+data class Position(var position: Int) {
+    fun move() {
+        position++
+    }
+}
```
position 프로퍼티가 public var로 외부에 노출되어 있는데, 
이렇게 노출되어 있다면 어떤 문제점이 있을까요 ?

namonak replied:
`position` 프로퍼티를 외부로 노출 시키면, `Position` 클래스를 정의한 것이 의미가 없을것 같다는 생각이 듭니다. 😢 
그래서 `position` 프로퍼티의 접근제한자를 `private`으로 변경하였는데, 제가 생각해낸 부분이 리뷰어님께서 말씀해 주신 문제점이 맞는지 궁금하네요 😭 

- https://github.com/next-step/kotlin-racingcar/commit/7881d51bb25a22cfc479bfe2402cb500aa4b3044

그리고, 수정하다 보니 결과값 출력을 위한 `ResultView` 클래스에서 `getPosition` 메서드에 바로 접근하는 형태인데 이렇게 해도 괜찮을까요? 중간에 DTO 객체를 거치지 않고, View 클래스에서 VO 객체의 값을 직접 얻어오는 것이 올바른 구조인지 아직 감이 안 잡히네요 😭 
기존에 `Cars` , `Car` 클래스를 DTO 객채로 활용하고 싶어도, getter/setter 메서드 외에 다른 기능 로직들이 있어서 활용하지 않았습니다! 😃

sah3122 replied:
변경가능한 상태의 변수가 외부로 노출된다면 상태값이 언제 변경되었는지 추적이 어렵고, 사이드 이펙트를 예상하기 힘들어집니다.
변경 불가능한 상태(val)를 가질수 없다면, setter 를 private으로 막아보세요.

namonak replied:
이 코멘트에 대한 내용을 너무 늦게 깨달았네요 😢 
일단 이번 과제에서는 `Position` 객체를 불변 객체로 변환하였는데요. - https://github.com/next-step/kotlin-racingcar/pull/831#discussion_r1021109642

알려주신 방법도 꼭 기억하고 있겠습니다! 감사합니다 😄

sah3122 reviewed (changes requested):
자동차 경주 구현 잘해주셨습니다 👍 
몇가지 고민거리를 남겨두었는데 확인 부탁드려요 🙏 
앞서 이야기 드린것 처럼 DM은 언제나 환영입니다 !

namonak committed:
feat: modify the position property of the Position class to have an integer value greater than or equal to 0

namonak committed:
refactor: add an InputResultDto class for passing the input result

The following have been implement and test:
 - add a NumberOfCars class for entering the number of cars
 - add a PlayTime class to input the number of games
 - delete the Pair object from the InputView class
 - change the setInputData method name to InputData

namonak committed:
refactor: replace the readLine() method in the InputView class with the readln() method

namonak committed:
refactor: update the Postion class

The following have been implement and test:
 - add private visibility modifiers to the position property
 - add a getPosition() method that returns a position value

namonak committed:
refactor: add a MovingStrategy interface for movement strategies

namonak committed:
refactor: remove the duplicate declaration of the List<Car> collection property

sah3122 commented on src/main/kotlin/step3/input/InputResultDto.kt:
```diff
@@ -0,0 +1,14 @@
+import step3.input.value.NumberOfCars
+import step3.input.value.PlayTime
+
+class InputResultDto(private val numberOfCars: NumberOfCars, private val playTime: PlayTime) {
+    fun getNumberOfCars(): Int {
+        return numberOfCars.getNumberOfCars()
+    }
+
+    fun getPlayTime(): Int {
+        return playTime.getPlayTime()
+    }
+}
```
data class 로 정의하는것도 좋을것 같아요 😄

namonak replied:
그렇네요! 코멘트 남겨주신대로 `InputResultDto` 클래스를 data class로 정의하였습니다! 😃 

- https://github.com/next-step/kotlin-racingcar/commit/38aa8eaae64687718f67f36e67c85faed96986bb

sah3122 commented on src/main/kotlin/step3/input/InputResultDto.kt:
```diff
@@ -0,0 +1,14 @@
+import step3.input.value.PlayTime
+
+class InputResultDto(private val numberOfCars: NumberOfCars, private val playTime: PlayTime) {
+    fun getNumberOfCars(): Int {
```
코틀린에서는 property의 getter 오버라이딩을 지원합니다. 

https://kotlinlang.org/docs/properties.html#getters-and-setters

namonak replied:
코멘트 주신 내용을 아래와 같이 적용해 보려고 하는데요. 괜찮은 방법일까요? 그리고 `NumberOfCars` , `PlayTime` 객체의 프로퍼티명으로 `_` 를 추가하는게 올바른 네이밍 인지 모르겠습니다 😢 

```kotlin
data class InputResultDto(private val _NumberOfCars: NumberOfCars, private val _playTime: PlayTime) {
    val numberOfCars get() = _NumberOfCars.getNumberOfCars()
    val playTime get() = _playTime.getPlayTime()
}
```

dean-wev replied:
코틀린의 관행으로 언더스코어를 변수명 앞에 사용하는건 사용하지 않은 변수를 나타내는 것 또는 backing property 를 선언하는 경우 사용하는것으로 알소 있어요. 
현재는 언더스코어를 제거해도될것 같아요 ! 
다른 방법으로는 data class 의 프로퍼티를 NumberOfCars.getNumberOfCars(), playTime.getPlayTime() 로 선언하는것도 방법이 되겠네요 ㅎㅎ

namonak replied:
세세한 조언 감사합니다! 😄 
proverty의 getter 오버라이딩을 사용하니 불필요한 getter 메서드가 없어지면서 코드가 더 간결해지고 가독성이 높은 코드가 되었습니다.

- https://github.com/next-step/kotlin-racingcar/commit/b42bab754337597f78151159b7a9f4014274b9ba

sah3122 commented on src/main/kotlin/step3/domain/car/Cars.kt:
```diff
@@ -1,17 +1,15 @@
 
     fun play() {
-        cars.forEach { it.move(Utils.generateRandomNumber(Utils.RANDOM_NUMBER_START_INDEX, Utils.RANDOM_NUMBER_END_INDEX)) }
+        cars.forEach { it.move(MovingStrategyImpl()) }
```
현재 Cars의 play는 항상 동일한 전략만 사용할수 있네요.  
전략 패턴에 사용되는 객체는 외부에서 주입받는것도 좋은 방법이라고 생각해요.

namonak replied:
DM으로 귀찮게 해드렸는데, 잘 설명해 주셔서 감사합니다. 😄 
조언해 주신 대로 수정해 보았습니다. 😃 

- https://github.com/next-step/kotlin-racingcar/commit/e75c98ea97a13dbb36254e5fc2646432f06b5961

sah3122 commented on src/main/kotlin/step3/domain/car/action/MovingStrategy.kt:
```diff
@@ -0,0 +1,5 @@
+package step3.domain.car.action
+
+interface MovingStrategy {
```
코틀린은 fun interface도 지원합니다. 😄

namonak replied:
이건 진짜 너무 꿀팁이네요... `MovingStrategy` 인터페이스를 함수형 인터페이스(`fun interface`)로 변경하니, 테스트 코드와 프로덕션 코드가 모두 람다식으로 변경되어 코드가 한결 간결해 졌습니다. 😄 

- https://github.com/next-step/kotlin-racingcar/commit/c10017fe3af127ef008596ab815037d7a11805db

sah3122 commented on src/main/kotlin/step3/domain/car/action/MovingStrategyImpl.kt:
```diff
@@ -0,0 +1,14 @@
+
+private const val MOVING_MINIMUM_CONDITION_NUMBER = 4
+
+class MovingStrategyImpl : MovingStrategy {
```
MovingStrategyImpl 보다는 어떤 전략을 가지는 객체인지 클래스명으로 나타내는것이 어떨까요 ?

namonak replied:
코멘트 주신 대로 `MovingStrategyImpl` 클래스명을 `RandomMovingStrategy`로 변경하였습니다. 😃 

- https://github.com/next-step/kotlin-racingcar/commit/e33cd2a316b514db690fb00f254017ecc1cb9040

sah3122 commented on src/main/kotlin/step3/domain/car/action/MovingStrategyImpl.kt:
```diff
@@ -0,0 +1,14 @@
+
+class MovingStrategyImpl : MovingStrategy {
+    override fun isMovable(): Boolean {
+        if (MOVING_MINIMUM_CONDITION_NUMBER <= Utils.generateRandomNumber(Utils.RANDOM_NUMBER_START_INDEX, Utils.RANDOM_NUMBER_END_INDEX)) {
```
Util 클래스를 사용하지 않고 구현해주신 구현체를 랜덤 값을 사용하는 전략으로 정의해보세요 😄

namonak replied:
조언해 주신대로, `Utils` 클래스는 사용하지 않고, `RandomMovingStrategy` 클래스에서 랜덤 값을 사용하는 로직을 처리하도록 리팩토링 하였습니다.

- https://github.com/next-step/kotlin-racingcar/commit/c10017fe3af127ef008596ab815037d7a11805db

sah3122 commented on src/main/kotlin/step3/domain/car/action/MovingStrategyImpl.kt:
```diff
@@ -0,0 +1,14 @@
+
+class MovingStrategyImpl : MovingStrategy {
+    override fun isMovable(): Boolean {
+        if (MOVING_MINIMUM_CONDITION_NUMBER <= Utils.generateRandomNumber(Utils.RANDOM_NUMBER_START_INDEX, Utils.RANDOM_NUMBER_END_INDEX)) {
+            return true
+        }
+        return false
```
```suggestion
        return if (MOVING_MINIMUM_CONDITION_NUMBER <= Utils.generateRandomNumber(Utils.RANDOM_NUMBER_START_INDEX, Utils.RANDOM_NUMBER_END_INDEX)) {
           true
        } else {
           false
        }
```

코틀린에서 if문은 표현식으로도 사용되니 이렇게 표현할수도 있겠네요 😄

namonak replied:
@sah3122 리뷰어님, 조언 감사합니다 😄 

한가지 궁금한 점이 있는데요. 조언해 주신 대로 하는 것은 [객체 지향 생활 체조 원칙](https://edu.nextstep.camp/s/56sRKVkV/ls/6RSdcCIN)의 `2. else 예약어를 쓰지 않는다.` 와는 무관한 것인가요?

namonak replied:
@sah3122 @dean-wev 리뷰어님,

해당 기능이 람다식으로 리팩토링되어, 조언해 주신 방법대로 수정을 못하게 되었습니다 😭 
알려주신 내용은 꼭 기억해 두고 추후에 사용하겠습니다 😄 

- https://github.com/next-step/kotlin-racingcar/commit/c10017fe3af127ef008596ab815037d7a11805db

sah3122 commented on src/main/kotlin/step3/domain/car/status/Position.kt:
```diff
@@ -1,7 +1,15 @@
 package step3.domain.car.status
 
-data class Position(var position: Int) {
+data class Position(private var position: Int) {
```
DM 으로 알려드린것 처럼 data class 를 불변 객체로 사용하는 방법도 경험해보세요 ! 
https://meetup.toast.com/posts/217

namonak replied:
@sah3122 @dean-wev 리뷰어님,

코멘트 주신 내용에 대해서 질문이 하나 있습니다! `Position` 클래스의 `position` 프로퍼티를 `val`로 선언하는 경우, 불변 객체로 사용은 가능할것 같은데요. 대신 `position` 프로퍼티 값을 증가시키기 위해서 있었던, `move()` 메서드는 불필요한 메서드가 되어 버릴것 같고, `Car` 객체에서는 위치 정보가 변경될 때마다, `Position` 객체를 계속 생성해야 할것 같은데... 이래도 괜찮은 걸까요? 😢 일단 리뷰어님이 조언해 주신대로 수정 진행해 보겠습니다 😄

namonak replied:
코멘트 남겨주신대로 `Position` 객체를 불변 객체로 변경해 보았습니다 😄 

- https://github.com/next-step/kotlin-racingcar/commit/e93c73c5da7fae7e527be4430860a57599460af1

namonak replied:
아... https://github.com/next-step/kotlin-racingcar/pull/831#discussion_r1020909865 코멘트에서 이미 대안을 주셨네요. 😭 

> 변경 불가능한 상태(val)를 가질수 없다면, setter 를 private으로 막아보세요.

좋은 깨달음 얻고 갑니다 😃

sah3122 commented on src/main/kotlin/step3/domain/car/status/Position.kt:
```diff
@@ -1,7 +1,15 @@
-data class Position(var position: Int) {
+data class Position(private var position: Int) {
+    init {
+        require(position >= 0) { "자동차의 위치는 0 이상이어야 합니다." }
```
매직넘버는 의미있는 상수로 정의하여 가독성을 높일수 있어요.

namonak replied:
예전과 동일한 실수를 했네요 😭 주의하겠습니다 🙏 
`Position` 클래스에 기본 위치 값을 정의한 `DEFAULT_POSITION` 상수 값을 추가하였습니다.

- https://github.com/next-step/kotlin-racingcar/commit/823f93f3ec93e027801381839135208cb40bab1f

sah3122 commented on src/main/kotlin/step3/input/value/NumberOfCars.kt:
```diff
@@ -0,0 +1,11 @@
+package step3.input.value
+
+data class NumberOfCars(private val numberOfCars: Int) {
```
val 로 정의한 변수는 외부에 노출되더라도 값을 변경할수 없으니 private 으로 정의하지 않아도 되지않을까요 ?

namonak replied:
불필요한 `private` 가시성 수정자를 삭제하니, 해당 프로퍼티와 관련된 불필요한 getter 메서드도 삭제되었습니다! 😄 

- https://github.com/next-step/kotlin-racingcar/commit/b42bab754337597f78151159b7a9f4014274b9ba

sah3122 reviewed (approved):
피드백 반영잘해주셨네요 👍 
몇가지 사소한 코멘트 남겨두었는데 확인 부탁드리며 다음 미션도 화이팅입니다 !
